jibby.go

```diff
@@ -0,0 +1,529 @@
+}
+
+// readAfterWS discards JSON white space and returns the next character.
+// Any error that occurs is returned without wrapping.
+func (d *Decoder) readAfterWS() (byte, error) {
+	var ch byte
+	var err error
+	for {
+		ch, err = d.json.ReadByte()
+		if err != nil {
+			// Don't use newReadError here as we don't know if there must be
+			// another character.  Let the caller decide.
+			return 0, err
+		}
+		switch ch {
+		case ' ', '\t', '\n', '\r':
+		default:
+			return ch, nil
+		}
+	}
+}
+
+// skipWS consumes white space but leaves the next character in the input
+// stream.  Any error that occurs is returned without wrapping.
+func (d *Decoder) skipWS() error {
+	_, err := d.readAfterWS()
+	if err != nil {
+		return err
+	}
+	_ = d.json.UnreadByte()
+	return nil
+}
+
+// readCharAfterWS reads a specific character after white space or errors
+// if the character is not available.  Any read error is returned, with
+// EOF upgraded to io.ErrUnexpectedEOF.
+func (d *Decoder) readCharAfterWS(b byte) error {
+	ch, err := d.readAfterWS()
+	if err != nil {
+		return newReadError(err)
+	}
+	if ch != b {
+		return d.parseError(ch, fmt.Sprintf("expecting '%c'", b))
+	}
+	return nil
+}
+
+// readNextChar expects a specific character to be next in the input stream and
+// errors otherwise.  Any read error is returned, with EOF upgraded to
+// io.ErrUnexpectedEOF.
+func (d *Decoder) readNextChar(b byte) error {
+	ch, err := d.json.ReadByte()
+	if err != nil {
+		return newReadError(err)
+	}
+	if ch != b {
+		return d.parseError(ch, fmt.Sprintf("expecting '%c'", b))
+	}
+	return nil
+}
+
+// readNameSeparator expects the ':' character after optional white space and
+// errors if it not found.  It handles other errors like readCharAfterWS.
+func (d *Decoder) readNameSeparator() error {
+	err := d.readCharAfterWS(':')
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// readNameSeparator expects the '}' character after optional white space and
+// errors if it is not found.  It handles other errors like readCharAfterWS.
+func (d *Decoder) readObjectTerminator() error {
+	err := d.readCharAfterWS('}')
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// readNameSeparator expects the '"' character after optional white space and
+// errors if it is not found.  It handles other errors like readCharAfterWS.
+func (d *Decoder) readQuoteStart() error {
+	err := d.readCharAfterWS('"')
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// readSpecificKey expects and consumes a specific series of bytes representing
+// an object key, and errors if it is not found.  The starting quote character
+// must have already been consumed from the input stream.  The closing quote and
+// subsequent name separator will also be consumed from the input stream.  It
+// handles other errors like readCharAfterWS.
+func (d *Decoder) readSpecificKey(expected []byte) error {
+	charsNeeded := len(expected) + 1
+	key, err := d.peekBoundedQuote(charsNeeded, charsNeeded)
```

Comment:
[nit] This is a purely a code layout question: you use `peekBoundedQuote` and this seems to be the only place it is used in this file. It appears _after_ all the numeric methods below. I think moving `peekBoundedQuote` below this method and above `peekNumber` would help with readability.